action/staking/stake.go

```diff
@@ -0,0 +1,163 @@
+		return false, errors.Wrap(action.ErrWrongTxType, err.Error())
+	}
+	err = action.ValidateBasic(tx.RawBytes(), st.Signers(), tx.Signatures)
+	if err != nil {
+		return false, err
+	}
+
+	err = action.ValidateFee(ctx.FeePool.GetOpt(), tx.Fee)
+	if err != nil {
+		return false, err
+	}
+
+	if len(st.StakeAddress) == 0 {
+		return false, action.ErrMissingData
+	}
+
+	if st.ValidatorAddress == nil {
+		return false, action.ErrMissingData
+	}
+
+	_, err = st.ValidatorPubKey.GetHandler()
+	if err != nil {
+		return false, action.ErrInvalidPubkey
+	}
+
+	coin := st.Stake.ToCoinWithBase(ctx.Currencies)
+	if !coin.IsValid() {
+		return false, errors.Wrap(action.ErrInvalidAmount, coin.String())
+	}
+
+	if coin.Currency.Name != "OLT" {
+		return false, action.ErrInvalidCurrency
+	}
+
+	_, ok := ctx.Currencies.GetCurrencyByName("OLT")
+	if !ok {
+		return false, action.ErrInvalidCurrency
+	}
+
+	err = ctx.Balances.CheckBalanceFromAddress(st.StakeAddress, coin)
+	if err != nil {
+		return false, action.ErrNotEnoughFund
+	}
+	return true, nil
+}
+
+func (s stakeTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing Apply stake Transaction for ProcessCheck", tx)
+	ok, result = runCheckStake(ctx, tx)
+	ctx.Logger.Debug("Result Apply stake Transaction for ProcessCheck", ok, result)
+	return
+}
+
+func (s stakeTx) ProcessDeliver(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing Apply stake Transaction for ProcessDeliver", tx)
+	ok, result = runCheckStake(ctx, tx)
+	ctx.Logger.Debug("Result Apply stake Transaction for ProcessDeliver", ok, result)
+	return
+}
+
+func (s stakeTx) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	ctx.Logger.Debug("Processing Apply stake Transaction for ProcessFee", signedTx)
+	return action.BasicFeeHandling(ctx, signedTx, start, size, 2)
+}
+
+func runCheckStake(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	st := &Stake{}
+	err := st.Unmarshal(tx.Data)
+	if err != nil {
+		return false, action.Response{Log: err.Error()}
+	}
+
+	stake := identity.Stake{
+		ValidatorAddress: st.ValidatorAddress,
+		StakeAddress:     st.StakeAddress,
+		Pubkey:           st.ValidatorPubKey,
+		ECDSAPubKey:      st.ValidatorECDSAPubKey,
+		Name:             st.NodeName,
+		Amount:           st.Stake.Value,
+	}
+
+	coin := st.Stake.ToCoinWithBase(ctx.Currencies)
+
+	err = ctx.Balances.MinusFromAddress(st.StakeAddress, coin)
+	if err != nil {
+		return false, action.Response{Log: errors.Wrap(err, st.StakeAddress.String()).Error()}
+	}
+
+	err = ctx.Delegators.Stake(st.ValidatorAddress, st.StakeAddress, st.Stake.Value)
+	if err != nil {
+		return false, action.Response{Log: errors.Wrap(err, st.StakeAddress.String()).Error()}
+	}
+
+	err = ctx.Validators.HandleStake(stake)
+	if err != nil {
+		return false, action.Response{Log: err.Error()}
```

Comment:
We reorganized and unified errors and error codes since we introduced governance. In folder status_codes/codes.go, we defined the error codes needed in our transactions. In folder identity/, you can create a new file 'errors.go' and put all error objects there, like data/balance/errors.go

Reply:
Code is ok, but as a developer I want to understand directly where an error (even when I use API). In case of HandleUnstake it manipulates with 3 different keys, and different errors could take a place. As a normal user, he must not receive an error in this place (only when he will try something to hack with storage)